cleanup embedder/compositor/constellation/script messages

paulrouget opened the pull request:
Fix: #17226 #17200 #17201

This is work in progress. Some tests still fail.
I'd like to get early feedback as it's a pretty large PR.

There is nothing fundamentally new. Basically, I added TopLevelBrowsingContrextId to the relevant messages between the embedder, the compositor and the constellation, and enforced the PipelineId to be attached to each ScriptMsg (see #17201).

I unaliased all the ScriptMsg. It was getting difficult to understand the nature of the message as ScriptMsg was used aliased CompositorMsg sometimes (CompositorMsg is an actually type of message already). I renamed constellation_chan to script_to_constellation_chan, again, for clarification.

This cleanup code is necessary for #15934 and for tabs support.

/cc @asajeffrey can I ask you to look at this? No need for a formal review, I need feedback at this stage.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17425)
<!-- Reviewable:end -->

paulrouget committed:
introduce ScriptToConstellationChan

paulrouget committed:
rename Browser to Servo as we support multiple browsers

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/pipeline.rs, components/constellation/constellation.rs
 * @kichjang: components/script/dom/dissimilaroriginwindow.rs, components/script/dom/workletglobalscope.rs, components/script/dom/window.rs, components/script_traits/lib.rs, components/script_traits/lib.rs and 18 more
 * @fitzgen: components/script/dom/dissimilaroriginwindow.rs, components/script/dom/workletglobalscope.rs, components/script/dom/window.rs, components/script_traits/lib.rs, components/script_traits/lib.rs and 18 more
 * @emilio: components/script/dom/webglrenderingcontext.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

paulrouget commented:
/cc @gterzian

asajeffrey commented on components/compositing/compositor.rs:
```diff
@@ -831,16 +822,39 @@ impl<Window: WindowMethods> IOCompositor<Window> {
-                self.on_key_event(ch, key, state, modifiers);
+                // FIXME(paul): this should be a message sent from embedder
+                // Steal a few key events for webrender debug options.
+                if modifiers.contains(CONTROL) && state == KeyState::Pressed {
```
It would be really nice if we didn't need this sort of hand-crafted code in servo itself, can we find a way to move it into the embedding app?

paulrouget replied:
Will happen later. Filed #17647

asajeffrey commented on components/constellation/constellation.rs:
```diff
@@ -799,17 +802,16 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
     fn handle_request(&mut self) {
         enum Request {
-            Script(FromScriptMsg),
+            Script((PipelineId, FromScriptMsg)),
```
Nit: `Script(PipelineId, FromScriptMsg)` rather than `Script((PipelineId, FromScriptMsg))`?

paulrouget replied:
This is not consistent with `NetworkListener`. Also, it's going to make the following select map a bit more complex.

asajeffrey reviewed (commented):
Yay, this looks a lot better! The addition of `TopLevelBrowserId`s seems to have cleaned the code up a lot!

paulrouget commented:
Thank you. I'll go ahead and finalize this PR and then ask for a formal review.

paulrouget commented:
@gterzian can you give me an informal review of this PR? I'd like to get your input before asking for a review to Alan.

paulrouget commented:
I still need to figure out what to do in `handle_load_start_msg`.

Also, I'd very much like rename `top_level_browsing_context_id` to `browser_id`.

gterzian commented:
@paulrouget Ok here are my comments. 

Have you considered:

1. Replacing the 'ad hoc' `Window.set_browser_id` with a new `WindowMethod`, that would also reflect the fact that new browsers could be created on an ongoing basis(for example if the JS calls `Window.open`), maybe something like `new_browser_created(browser_id)`?
2. Adding a new `WindowEvent::SelectBrowser(browser_id)`? 
3. Adding a new `WindowEvent::NewBrowser(target_url)`. 

Those proposals could seen as candidates to replace the current dance taking place in `servo/main`, simply by making `WindowEvent::NewBrowser(target_url)` the initial event that is sent through(replacing `vec![WindowEvent::InitializeCompositing]` with `vec![WindowEvent::NewBrowser(target_url)]`), to which Servo could respond by calling it's own `create_browser` method,  followed by calling `self.compositor.initialize_compositing()` as well as `self.compositor.window.new_browser_created(browser_id)` once constellation has responded with a browser_id(You could even go full circle by not showing anything until window sends a `WindowEvent::SelectBrowser(browser_id)` in response to the `new_browser_created` call, after having checked internally that this was indeed the first browser that was created and none are already "selected")?

By the way this is a good example of why it could be useful for Servo to handle `WindowEvents` directly, as opposed to just passing those on to compositor. So the above changes should perhaps instead be made in https://github.com/servo/servo/pull/17269 after this one has merged...

paulrouget commented:
> Adding a new WindowEvent::SelectBrowser(browser_id)?

Agreed.

> Adding a new WindowEvent::NewBrowser(target_url).

Agreed.

> Replacing the 'ad hoc' Window.set_browser_id with a new WindowMethod, that would also reflect the fact that new browsers could be created on an ongoing basis(for example if the JS calls Window.open), maybe something like new_browser_created(browser_id)?

I don't think we want servo to have the ability to create new browsers. What we want is Servo to tell the embedder to handle a new url, and then up to the embedder to create a new browser.

To get the browser_id back, we would use this mechanism instead:

`WindowEvent::NewBrowser(ServoUrl, IpcSender<BrowserId>)`

What do you think?

> Those proposals could seen as candidates to replace the current dance taking place in servo/main, simply by making WindowEvent::NewBrowser(target_url) the initial event that is sent through(replacing vec![WindowEvent::InitializeCompositing] with vec![WindowEvent::NewBrowser(target_url)])

Ok

> to which Servo could respond by calling it's own create_browser method, followed by calling self.compositor.initialize_compositing()

This should actually disappear. `initialize_compositing` does nothing.

> as well as self.compositor.window.new_browser_created(browser_id) once constellation has responded with a browser_id

Let's not use a callback for this but an IpcSender as I suggested earlier. Because we want to know which browser_id correspond to which NewBrowser request.

Think about this case:

```rust
send(NewBrowser(url1));
send(NewBrowser(url2));

…

fn new_browser_created(browser_id) {
// here we don't know which browser_id correspond to request number 1 and request number 2.
}
```

> (You could even go full circle by not showing anything until window sends a WindowEvent::SelectBrowser(browser_id) in response to the new_browser_created call, after having checked internally that this was indeed the first browser that was created and none are already "selected")?

I think that's already the case, isn't it?

> By the way this is a good example of why it could be useful for Servo to handle WindowEvents directly, as opposed to just passing those on to compositor. So the above changes should perhaps instead be made in #17269 after this one has merged...

I think we should at least use WindowEvents instead of new `Servo::` methods. It makes more sense.

gterzian commented:
@paulrouget 

> To get the browser_id back, we would use this mechanism instead: WindowEvent::NewBrowser(ServoUrl, IpcSender<BrowserId>)

Yes makes sense.

> I don't think we want servo to have the ability to create new browsers. What we want is Servo to tell the embedder to handle a new url, and then up to the embedder to create a new browser.

Ok I see, so servo would ask the embedder to "handle a url", to which the embedder could respond by sending a `WindowEvent::NewBrowser(ServoUrl, IpcSender<BrowserId>)`...

> not showing anything until window sends a WindowEvent::SelectBrowser(browser_id). I think that's already the case, isn't it?

Yes it's the case, although it's servo that calls it's own `select_browser` method, so in the light of the above it could be instead servo handling a `WindowEvent::SelectBrowser(browser_id)` event. (So would require some new code in the glutin window, in response to receiving on a `IpcSender<BrowserId>`.

paulrouget commented:
@gterzian Can you take a look at the 2 last commits?

Using `WindowEvent` instead of methods is more consistent, but it's cumbersome. I believe we should keep it that way for now for consistency, but eventually, we might want to move from messages to methods.

gterzian commented:
I see. 

Perhaps some of the cumbersomeness can be removed later by finding a way to get rid of the initial `handle_events(vec![WindowEvent::NewBrowser(target_url, sender)]);`, and simply have servo handle incoming events, where window would, as part of it's own initialization, fire a `WindowEvent::NewBrowser(target_url, sender)`. In fact one could get rid of `window.set_browser_id(browser_id);`, since window could internally receive on the `sender`, and simply store the `browser_id` by itself...

gterzian commented:
Maybe one way to get rid of the initial `vec![WindowEvent::NewBrowser(target_url, sender)]`, is by adding a `open_new_browser(target_url` WindowMethod, which could be called in `servo/main`, to which window could respond by firing the initial WindowEvent::NewBrowser(target_url, sender) (and storing the `browser_id` after having received it from the `sender`, which window could receive on internally, since it would be firing the event)

gterzian commented:
I guess `maybe_open_new_browser(target_url)` is a better name, since it would be up to the embedder to allow it or not...

paulrouget commented:
We'll see later about the cosmetic changes. As for now, I'll keep your initial suggestion and use an event. For consistency.

Later, we will take the time to re-consider the decision to use messages for embedder -> libservo communication, and callbacks for libservo -> embedder communication.

paulrouget commented:
Depends on #17709, #17686 and #17685.

paulrouget committed:
Ability to create new TopLevelBrowsingContext

paulrouget committed:
Use BrowserId and rename Browser to Servo in ports/

paulrouget committed:
Event to make a browser visible

paulrouget committed:
Add TopLevelBrowsingContextId to messages between Embedder <-> Compositor <-> Constellation

paulrouget committed:
make use of ScriptToConstellationChan

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -1823,7 +1836,14 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+        if pipeline_is_top_level_pipeline {
+            // Notify embedder top level document finished loading.
+            self.compositor_proxy.send(ToCompositorMsg::LoadComplete(top_level_browsing_context_id));
+        }
```
`handle_load_complete_msg` and `handle_load_start_msg` use 2 different ways of checking if the pipeline is the top level pipeline. load_start checks if there's a parent_info, load_complete checks if the browsing_context's pipeline is the same as the pipeline argument.

These 2 ways are both valid, right?

If so, which would is the most sensible?

asajeffrey replied:
The difference IIRC is how mozbrowser iframes are treated. Once mozbrowser iframes are removed we can just check parent_info, but for the moment it does make a difference which one we do.

paulrouget commented:
@asajeffrey This is ready for review.

paulrouget commented:
A summary of this PR:

- messages between constellation/compositor and embedder include the top level browsing context id when it makes sense
- messages from script automatically come with a pipeline id, which will automatically give us the browsing context id of each of these messages
- 2 new `WindowEvent` to create a top level browsing context (aka "browser") and toggle its visibility
- removed as much as possible of browsing context logic code from compositor and moved it to constellation. This is actually the goal of this PR. Once this lands, the next step will be to short circuit the compositor step when the embedder communicate with the constellation.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17735) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17775) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #17499) made this pull request unmergeable. Please resolve the merge conflicts.

paulrouget commented:
@asajeffrey I rebased. This PR touches many files and easily get unmergeable. Can you please review even if it needs a rebase?

asajeffrey commented on components/script/dom/workletglobalscope.rs:
```diff
@@ -68,14 +69,17 @@ impl WorkletGlobalScope {
+                                                    ScriptToConstellationChan {
+                                                        sender: init.to_constellation_sender.clone(),
+                                                        pipeline_id: pipeline_id,
+                                                    },
```
Might be worth let-binding this value.

asajeffrey commented on components/script/script_thread.rs:
```diff
@@ -2001,7 +2001,10 @@ impl ScriptThread {
+                                 ScriptToConstellationChan {
+                                     sender: self.script_sender.clone(),
+                                     pipeline_id: incomplete.pipeline_id,
+                                 },
```
Let-bind this?

asajeffrey commented on ports/glutin/window.rs:
```diff
@@ -182,6 +182,8 @@ pub struct Window {
     mouse_down_point: Cell<Point2D<i32>>,
     event_queue: RefCell<Vec<WindowEvent>>,
 
+    browser_id: Cell<Option<BrowserId>>,
```
Why is this an Option? Is it because of initalization? If so, can we add a comment to that effect?

paulrouget replied:
It is because of initalization. Adding a comment.

asajeffrey reviewed (commented):
This looks very nice! No huge blockers, just some nits.

paulrouget commented on components/compositing/compositor.rs:
```diff
@@ -475,55 +473,56 @@ impl<Window: WindowMethods> IOCompositor<Window> {
-            (Msg::LoadStart, ShutdownState::NotShuttingDown) => {
-                self.window.load_start();
+            (Msg::LoadStart(top_level_browsing_context_id), ShutdownState::NotShuttingDown) => {
+                self.window.load_start(top_level_browsing_context_id);
```
For both load_start and load_complete, here we are talking about the "Tab". It is basically used to spin the loading throbber in the user interface.

We don't want to be notified of the loading status of all the pipelines.

asajeffrey replied:
I'm concerned about messages arriving in weird orders, something like:
```
1. Compositor -> Constellation: Load A in tab T
2. Compositor -> Constellation: Load B in tab T
3. Constellation -> Compositor: Loaded A in tab T
4. Constellation -> Compositor: Loaded B in tab T
```
The problem being that the constellation and the compositor see messages 2 and 3 in different orders, so may disagree about which document finished loading.

paulrouget replied:
I'm not sure to understand how that could be a problem. We fire LoadComplete only if the document that has finished loading is also the current top level pipeline of the browsing context. Also, with this PR, the compositor doesn't need to agree on anything, as all this logic is in the constellation now.

Things would get mixed up if the current browsing context top level pipeline is not updated at the right time, but that would not be related to my work I think.

paulrouget replied:
I just realized that you're talking about the embedder when you mention compositor. You're saying that if we don't use the pipeline_id, the embedder doesn't know which document has loaded. This is true.

From the embedder perspective, it's not an information we need at this point. We want to know the loading status of the tab, and what is the current URL, which all can be precisely inferred from the `HeadParsed` and `HistoryChanged` messages. 

Also - we don't have a way to identify pipelines/documents in the embedder yet. Maybe if we end up exposing `PipelineId` later, we will add pipeline_id to the `HeadParsed` and `LoadComplete` messages.

asajeffrey replied:
Yes, sorry I meant embedder.

I don't understand how the embedder is meant to handle the case where they open document A then B in the same tab. When a load complete message arrives, they have no way to know whether it was A or B that completed. Is the embedder expected to keep a counter of currently loading documents, and (e.g.) stop the spinner when the count reaches zero?

paulrouget replied:
irc discussion: http://logs.glob.uno/?c=mozilla%23servo#c727681

ok - let me try to describe the problem you're describing, to make sure we are on the same page:

- Load A happens, LoadStart is fired
- Load B happens right after, LoadStart is fired
- A finishes to load
- B finishes to load

We don't want to be in a situation where we fire LoadComplete when A finished loading, because in that case, the embedder would not expect another LoadComplete and would assume it's B that finished loading.

My assumption is that we can only end up in these situations:

```
> Load A
> Load B
< LoadStart A
< LoadComplete A
< LoadStart B
< LoadComplete B
```

or

```
> Load A
> Load B
< LoadStart A
< LoadStart B
< LoadComplete B
```

What we don't want is

```
> Load A
> Load B
< LoadStart A
< LoadStart B
< LoadComplete A
< LoadComplete B
```

The intended logic of the code is to only fire LoadComplete for X if X is the current pipeline.

So first, this is wrong, because we can be in a situation where A is still the current pipeline because B is still pending. So my PR is wrong here. We also need to check the pending changes and make sure nothing else is expected.

Next, can message latency be an issue? Which is I think what you were getting at. Is it possible to get LoadComplete for A inside the constellation and not know that embedder requested to load B? Yes. But, what really matters is to not fire LoadComplete A after we fired **LoadStart** B. The embedder relies on the LoadStart message, not on when it calls Load A / Load B.

And LoadStart and LoadComplete are states known by the constellation. There is not reason for this to have any race here.

The only real issue would be if we get messages unordered, which I hope is not the case :)

Conclusion: I need to check for pending changes before firing LoadComplete. And I believe the asynchronous nature of the messages is not a problem as long as the embedder relies on LoadStart.

I hope this clarify the situation.

Let me know.

asajeffrey replied:
Ah, so the idea is:

a) The compositor always fires a `LoadStart` for every load requested.
b) The comopositor only fires a `LoadComplete` when the most recent `LoadStart`ed piepline finishes.

Thia means that the compositors spinner logic can be to keep a counter, incremented when it sends a `Load`, decremented when it receives a `LoadStart`. The spinner is started when a `Load` is sent, and stopped when a `LoadComplete` is received and the counter is zero.

So yes, the embedder can implement a spinner, though I think they do need a counter to do it.

Are we now in sync?

paulrouget replied:
a) and b), yes.

> The spinner is started when a Load is sent

Yes, or when it receives LoadStart.

>  and stopped when a LoadComplete is received and the counter is zero.

If LoadComplete is received, that means the latest Load has finished loading. Spinner can be stopped.

No need to keep a counter because where would be these other loading documents? Aren't they past history? As in, you would only reach them by clicking the back button.

If we do Load A then Load B, once B has loaded, even if A it's still loading, it's not the current pipeline of the browsing context anymore, so the spinner should not be spinning.

paulrouget commented on components/compositing/windowing.rs:
```diff
@@ -62,15 +62,20 @@ pub enum WindowEvent {
     ToggleWebRenderProfiler,
+    Reload(TopLevelBrowsingContextId),
+    /// Create a new top level browsing context
+    NewBrowser(ServoUrl, IpcSender<TopLevelBrowsingContextId>),
```
Not yet. Ok with follow-up? I think this requires some more work in the constellation.

asajeffrey replied:
Sure.

paulrouget replied:
https://github.com/servo/servo/issues/18006 I'll take care of that later.

paulrouget commented on components/constellation/constellation.rs:
```diff
@@ -263,6 +266,7 @@ pub struct Constellation<Message, LTF, STF> {
     pending_changes: Vec<SessionHistoryChange>,
 
     /// The currently focused pipeline for key events.
+    /// FIXME(paul): there should be one per top level browsing context.
```
I'll remove the comment. I believe this is necessary. This can be discussed here: https://github.com/servo/servo/issues/17401

paulrouget commented:
@asajeffrey addressed your comments.

asajeffrey reviewed (commented):
No blockers here.

paulrouget commented:
@asajeffrey I added an extra commit. This is to make sure that we don't send LoadComplete if there are pending changes that will replace the current document.

our discussion earlier: http://logs.glob.uno/?c=mozilla%23servo#c728682

I hope I addressed your concerns. If there's anything you're not clear about, let me know.

asajeffrey commented on components/constellation/constellation.rs:
```diff
@@ -1880,13 +1880,31 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+                            warn!("No pipeline associated with pipeline_id {:?}", change.new_pipeline_id);
+                            false
+                        }
+                });
```
It's not obvious to me why you need this extra check rather than just `self.pending_changes.iter().any(|change| change.browsing_context_id == top_level_browsing_context_id)`. All the changes are for the top-level browsing context, so they'll automatically have no parent, won't they?

asajeffrey reviewed (commented):
I'm off on PTO next week, you can r=me on this, I'll let you to decide how to deal with the one last comment I made.

paulrouget commented:
You're right. I can't see a scenario where this will be true. About to remove that test and r+ the PR.

canaltinova commented:
@bors-servo delegate+

bors-servo commented:
:v: @paulrouget can now approve this pull request

paulrouget commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit a8ebe0f has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey a8ebe0f914570a7ccaead33a4774625ad3d678a9 -->

bors-servo commented:
:hourglass: Testing commit a8ebe0f914570a7ccaead33a4774625ad3d678a9 with merge 76700a3aea0a25ecbd5adbf8974e49ad60271fbc...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5325)

jdm commented:
```
  ▶ TIMEOUT [expected OK] /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation_by_user_activation_without_user_gesture.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-iframe-element/iframe_sandbox_allow_top_navigation_by_user_activation_without_user_gesture.html:
  │ TIMEOUT [expected PASS] The sandboxed iframe should post a message saying the test was in the state of 'PASS'.
  └   → Test timed out
```

paulrouget commented:
I made 2 changes:

- better commit messages (I'm wondering if it's better to squash everything…)
- fixed the test (wrong pipeline_id was passed along the PostMessage message

paulrouget commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit 10b5266 has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 10b52669d7dbe0f7d538bbd51135b66ad9488890 -->

bors-servo commented:
:hourglass: Testing commit 10b52669d7dbe0f7d538bbd51135b66ad9488890 with merge 74558990b258cb55f230ebe8ec3fc557fd286f94...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8310), [arm32](http://build.servo.org/builders/arm32/builds/8532), [arm64](http://build.servo.org/builders/arm64/builds/8565), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8651), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5417), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5452), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8648), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1126), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1127), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5350), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5408), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1128), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1132), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3893)
Approved by: asajeffrey
Pushing 74558990b258cb55f230ebe8ec3fc557fd286f94 to master...
